lib/Sema/CSSimplify.cpp

```diff
@@ -5228,6 +5228,107 @@ Type ConstraintSystem::simplifyAppliedOverloads(
   return fnType;
 }
 
+/// Returns the `call` declarations (if they exist) implemented or inherited
+/// by a type.
+/// This function may be slow for deep class hierarchies and multiple protocol
+/// conformances, but it is invoked only after other constraint simplification
+/// rules fail.
+static llvm::DenseSet<CallDecl *>
+getCallDeclarations(Type type, ConstraintSystem &CS,
+                    const ConstraintLocatorBuilder &locator) {
+  auto canType = type->getCanonicalType();
+  auto it = CS.CallDeclCache.find(canType);
+  if (it != CS.CallDeclCache.end()) return it->second;
+
+  // Calculate `call` declarations for composite types with multiple components
+  // (protocol composition types and archetypes).
+  auto calculateForComponentTypes =
+      [&](ArrayRef<Type> componentTypes) -> llvm::DenseSet<CallDecl *> {
+    llvm::DenseSet<CallDecl *> callDecls;
+    for (auto componentType : componentTypes) {
+      auto *nominal = componentType->getAnyNominal();
+      callDecls.insert(nominal->getCallDeclarations().begin(),
+                       nominal->getCallDeclarations().end());
+    }
+    return callDecls;
+  };
+
+  // Calculate `call` declarations.
+  auto calculate = [&]() -> llvm::DenseSet<CallDecl *> {
+    // If this is an archetype type, check if any types it conforms to
+    // (superclass or protocols) have`call` declarations.
+    if (auto archetype = dyn_cast<ArchetypeType>(canType)) {
+      SmallVector<Type, 2> componentTypes;
+      for (auto protocolDecl : archetype->getConformsTo())
+        componentTypes.push_back(protocolDecl->getDeclaredType());
+      if (auto superclass = archetype->getSuperclass())
+        componentTypes.push_back(superclass);
+      return calculateForComponentTypes(componentTypes);
+    }
+
+    // If this is a protocol composition, check if any of its members have
+    // `call` declarations.
+    if (auto protocolComp = dyn_cast<ProtocolCompositionType>(canType))
+      return calculateForComponentTypes(protocolComp->getMembers());
+
+    // Otherwise, this must be a nominal type.
+    // Structural types cannot define `call` declarations.
+    auto nominal = canType->getAnyNominal();
+    if (!nominal) return llvm::DenseSet<CallDecl *>();
+
+    // If this type conforms to protocols, look up all of their `call`
+    // declaration requirements.
+    // FIXME: Fix this to handle overridden requirements, e.g. using
```

Comment:
TODO/WIP: This logic for "getting all call declarations" doesn't work properly.
It should mimic/be unified with `lookupQualified`, as mentioned in [this post](https://forums.swift.org/t/question-what-should-calldecl-inherit-from/21815/9?u=dan-zheng) by @slavapestov.

Reply:
Done in https://github.com/apple/swift/pull/23517/commits/94a384feb522c132c6ec2e64f50daa53c940683a.

To find all `call` declarations, `TC.lookupMember` is used, which internally calls `lookupQualified`. This approach can be back-ported to `@dynamicCallable` and `@dynamicMemberLookup` to reduce code dupe (and improve correctness, in the case of `@dynamicCallable`). I'm interested in doing that in a follow-up.